### DISCUSSION TITLE
Entry level IMGUI & CMAKE example in Ubuntu

### DISCUSSION THREAD

**Discussion by jochoalions** [ocornut/imgui#5970] [New Users Build/Link/Run issues ONLY!]
I am looking for an example of how to setup CMAKE to build the simple imgui-demo program. I have found a number of very good and easy-to-follow tutorials on Cmake and Imgui, but I cannot even get to the point of being able to compile and run the imgui-demo program as a standalone. I fear that I am over-complicating the issue, given the extraordinary amount of excellent documentation throughout the Imgui source code. But that is where I am at and would appreciate some pointers or some help.

Does anybody know of a repository I can look into, or can provide some pointers on how to setup the CMakeLists.txt files?  (Or any other configuration process for that matter, since I am new to CMake and Ubuntu, I can learn any other tool.)

Thank you kindly!!

PS: Or if someone is willing to coach me one-on-one, I would certainly pay for your assistance.

**Comment by ocornut** [comment]
The biggest hurdle in all the C++ ecosystem is compiling and linking with third-party libraries. In particular, it's a difficult problem in Windows, less so in Unix-style systems like Linux and OSX.

While Dear ImGui core files can easily compile pretty much anywhere, and are platform agnostic, the backends/ files are all associated to specific third-party libraries. It is specifically the compiling and linking of backends that can be tricky if you are not already using them.

My very personal stance is that providing cmakefile to build the examples applications (which showcase using ALL backends) is too difficult to do in a cross platform manner or would lead to problematic restrictions for some users. This is why I choose to not provide them. Many people disagree with this assumption, and in particular some have been maintaining good set of cmakefiles: https://github.com/ocornut/imgui/pull/1713
As it turns out, linking with third-party on Linux or OSX is easier than on Windows so it is likely you can use those files.

To fully answer your question (and I assume other people after me will) we need to know which backend you intend to use (e.g. GLFW + OpenGL, SDL + Vulkan..). If you already have an windowed app with 3d rendering running you would naturally pick backends matching the libraries your app is running, and it is generally in this situation that compiling examples is NOT a problem. If you don't have an app running already, I'd suggest using SDL+OpenGL or GLFW+OpenGL.

**Reply by jochoalions** [reply]
Thank you for such informative response. I am starting from scratch, and intend to do primarily 2D Agent-based Modelling & Simulations. Because all of the tutorials I have watch point towards GLFW+OpenGL, that is what I am considering using. I will take a good look at #1713, I appreciate the pointer.

**Reply by jochoalions** [reply]
@ocornut As far as you know, has anyone created CMake files for Linux/Ubuntu?  I am looking at  the files in #1713 and those are targeting Windows and Visual Studio, which I am not using; they are super helpful though, because I can see how he has structured the CMake files, but there is still a lot I do not understand or know.

So, specifically for **GLFW+OpenGL** on **Ubuntu**, I would like to find a working example for how to build the **imgui-demo** program using CMake, or find someone who would be willing to coach me to achieve that goal.

**Comment by sukesh-ak** [comment]
@jochoalions 
Here is an example you can start with - Dear ImGui with GLFW + OpenGL3 with cmake + vcpkg
[Tested with Windows & Linux]
https://github.com/sukesh-ak/ImGui-GLFW-OpenGL3

**Comment by connorjclark** [comment]
Here's some simple cmake:

```
FetchContent_Declare(imgui_external
	URL https://github.com/ocornut/imgui/archive/refs/tags/v1.90.8.tar.gz
	EXCLUDE_FROM_ALL
)
FetchContent_MakeAvailable(imgui_external)

add_library(imgui
	${imgui_external_SOURCE_DIR}/imgui.cpp
	${imgui_external_SOURCE_DIR}/imgui_draw.cpp
	${imgui_external_SOURCE_DIR}/imgui_tables.cpp
	${imgui_external_SOURCE_DIR}/imgui_widgets.cpp
	${imgui_external_SOURCE_DIR}/backends/imgui_impl_allegro5.cpp
)
target_include_directories(imgui PUBLIC ${imgui_external_SOURCE_DIR})
```

All you need to do is swap out the backend file, or use a different version tag.
